Python/fileutils.c

```diff
@@ -1814,7 +1823,14 @@ _Py_write_impl(int fd, const void *buf, size_t count, int gil_held)
             Py_BEGIN_ALLOW_THREADS
             errno = 0;
 #ifdef MS_WINDOWS
+            _doserrno = 0;
             n = write(fd, buf, (int)count);
+            int winerr = _doserrno;
+            if (n < 0 && errno == ENOSPC) {
+                /* write() on a non-blocking full pipe
+                   returns ENOSPC on Windows */
+                errno = EAGAIN;
+            }
```

Comment:
`_doserrno` isn't needed here. Also, this case isn't limited to just a full pipe. There are no partial writes up to the available space in non-blocking mode, contrary to the documentation. I tested this as far back as Windows NT 4.0. The official WinAPI documentation is wrong.

```suggestion
            // write() on a non-blocking pipe fails with ENOSPC on Windows if
            // the pipe lacks available space for the entire buffer.
            n = write(fd, buf, (int)count);
            if (n < 0 && errno == ENOSPC) {
                errno = EAGAIN;
            }
```